Source/Tests/LinuxSyscalls/SyscallsVMATracking.cpp

```diff
@@ -0,0 +1,367 @@
+
+        // Make new VMA with new flags, insert for length of range
+        auto NewOffset = OffsetDiff + Base;
+        auto NewLength = Top - Base;
+
+        auto Inserted =
+            VMAs.emplace(Base, VMAEntry{Current->Resource, Current, Current->ResourceNextVMA, Base, NewOffset, NewLength, NewOptions});
+        LOGMAN_THROW_A_FMT(Inserted.second, "VMA tracking error");
+        auto RestOfMapping = &Inserted.first->second;
+
+        if (Current->Resource) {
+          ListInsert(Current, RestOfMapping);
+        }
+
+        Current = RestOfMapping;
+      } else {
+        // Mapping starts in range, just change Prot
+        Current->Options.Prot = Prot;
+      }
+
+      if (HasTrailingPart) {
+        // ends after Range, split last part and insert with original flags
+
+        // Trim the mapping (possibly already trimmed)
+        Current->Length = Top - Current->Base;
+
+        // prot has already been changed
+
+        // Make new VMA with original flags, insert for remaining length
+        auto NewOffset = OffsetDiff + Top;
+        auto NewLength = MapTop - Top;
+
+        auto Inserted =
+            VMAs.emplace(Top, VMAEntry{Current->Resource, Current, Current->ResourceNextVMA, Top, NewOffset, NewLength, MapOptions});
+        LOGMAN_THROW_A_FMT(Inserted.second, "VMA tracking error");
+        auto TrailingMapping = &Inserted.first->second;
+
+        if (Current->Resource) {
+          ListInsert(Current, TrailingMapping);
+        }
+      }
+    }
+  }
+}
+
+// This matches the peculiarities algorithm used in linux ksys_shmdt (linux kernel 5.16, ipc/shm.c)
+uintptr_t SyscallHandler::VMATracking::ClearShmUnsafe(FEXCore::Context::Context *CTX, uintptr_t Base) {
+
+  // First first VMA at or after Base
+  // Iterate until first SHM VMA, with matching offset, get length
+  // Iterate for the rest of the length removing instances of found SHM
+
+  // returns first element that is greater or equal or ::end
+  auto Entry = VMAs.lower_bound(Base);
+
+  while (Entry != VMAs.end()) {
+    LOGMAN_THROW_A_FMT(Entry->second.Base >= Base, "VMA tracking corruption");
+    if (Entry->second.Base - Base == Entry->second.Offset && Entry->second.Resource &&
+        Entry->second.Resource->Iterator->first.dev == MRID_SHM) {
+
+      const auto ShmLength = Entry->second.Resource->Iterator->second.Length;
+      const auto Resource = Entry->second.Resource;
+
+      do {
+        if (Entry->second.Resource == Resource) {
+          if (ListRemove(&Entry->second)) {
+            if (Entry->second.Resource->AOTIRCacheEntry) {
+              FEXCore::Context::UnloadAOTIRCacheEntry(CTX, Entry->second.Resource->AOTIRCacheEntry);
+            }
+            MappedResources.erase(Entry->second.Resource->Iterator);
+          }
+          Entry = VMAs.erase(Entry);
+        } else {
+          Entry++;
+        }
+      } while (++Entry != VMAs.end() && (Entry->second.Base + Entry->second.Length - Base) <= ShmLength);
```

Comment:
`Entry` is already increased in the loop body (in both branches). Is the additional increment in the loop condition intended?

Reply:
Yes, the `++Entry` here is erroneous. Nice catch!